virtio-devices/src/mem.rs

```diff
@@ -736,30 +736,35 @@ impl EpollHelperHandler for MemEpollHandler {
                     EpollHelperError::HandleEvent(anyhow!("Failed to get resize event: {:?}", e))
                 })?;
 
-                let size = self.resize.size();
                 let mut config = self.config.lock().unwrap();
-                let mut signal_error = false;
-                let mut r = config.resize(size);
-                r = match r {
-                    Err(e) => Err(e),
-                    _ => match self.signal(VirtioInterruptType::Config) {
-                        Err(e) => {
-                            signal_error = true;
-                            Err(Error::ResizeTriggerFail(e))
-                        }
-                        _ => Ok(()),
-                    },
-                };
-                let ret_err = anyhow!("{:?}", r);
-                if let Err(e) = self.resize.send(r) {
-                    return Err(EpollHelperError::HandleEvent(anyhow!(
-                        "Sending \"resize\" response: {:?}",
-                        e
-                    )));
-                }
-                if signal_error {
-                    return Err(EpollHelperError::HandleEvent(anyhow!(ret_err)));
+                if let Err(e) = config.resize(self.resize.size()) {
+                    self.resize.send(Err(e)).map_err(|e| {
+                        EpollHelperError::HandleEvent(anyhow!(
+                            "Sending \"resize\" response: {:?}",
+                            e
+                        ))
+                    })?;
+                    return Ok(());
                 }
+
+                self.signal(VirtioInterruptType::Config).map_err(|e| {
+                    let e = Error::ResizeTriggerFail(e);
+                    let ret_err = anyhow!("{:?}", e);
+                    self.resize
+                        .send(Err(e))
+                        .map_err(|e| {
+                            EpollHelperError::HandleEvent(anyhow!(
+                                "Sending \"resize\" response: {:?}",
+                                e
+                            ))
+                        })
+                        .ok();
+                    EpollHelperError::HandleEvent(ret_err)
+                })?;
+
+                self.resize.send(Ok(())).map_err(|e| {
+                    EpollHelperError::HandleEvent(anyhow!("Sending \"resize\" response: {:?}", e))
+                })?;
```

Comment:
I think you didn't break anything, but TBH I'm not sure your version is simpler than the original one.

Reply:
The main refactoring was to separate error handling code from the successful code path using "map_err()?;". The current code looks quite easier (for me) to understand what is the successful code path, where it can go wrong and how each error is handled. 

The original code looks quite confusing to me, and that's why I started the refactoring. Of course, I am open to different opinions about how we can make it better.

Reply:
Also, the refactoring from `balloon` is following the same thought described above. Please also take a look. Thank you.